### PR TITLE
chore: refactoring of any-to-erc20 from erc20-fee-proxy, with test fixes

### PR DESCRIPTION
**Pull request by yomarion**
## Description of the changes

* I initially wanted to make a first step towards payment network implementation with inheritance. The scope for this first step is the payment network "any to erc20".
* Some small mistakes in the tests were made more obvious by this change, I fixed them.
* It also made visible that default networks can be a pitfall

Overall, if we trim ~350 lines from `fee-proxy-contract.ts`, that's a +300 / -511 (diff stats) refactoring with ease of reading, testing and maintenability.

## Current status

TLDR: ignore `fee-proxy-contract.ts`
* The file `fee-proxy-contract.ts` contains both the former way and the new one (350 lines are +/- duplicated). It's easy to keep the new one, but let's first validate the change on "any-to-erc20". 
* `yarn lint` fails for missing documentation, that I add if we agree on the principle

### CODE REVIEW

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 1]
```diff
@@ -48,397 +28,163 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
+    this.paymentNetworkId = ExtensionTypes.ID.PAYMENT_NETWORK_ANY_TO_ERC20_PROXY;
+  }
+
+  /**
+   * Creates the extensionsData to create the extension Any to ERC20
+   *
+   * @param creationParameters extensions parameters to create
+   *
+   * @returns IExtensionCreationAction the extensionsData to be stored in the request
+   */
```
Comment (yomarion) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:44]: I'm not sure the comments are useful on the child classes. Opinions welcome!

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 2]
```diff
@@ -48,397 +28,163 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
+    extensionAction: ExtensionTypes.IAction,
+  ): void {
+    const network =
+      extensionAction.parameters.network ||
+      request.extensions[this.paymentNetworkId]?.values.network;
+
+    // Nothing can be validated if the network has not been given yet
+    if (!network) {
+      return;
+    }
```
Comment (yomarion) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:168]: That's a (potentially big) functional change, but I think it's valid. 

The former logic was to give a default network (mainnet) but it makes a difference between mainnet & rinkeby behaviour.

File: packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts [hunk 3]
```diff
@@ -64,89 +64,92 @@ describe('extensions/payment-network/erc20/any-to-erc20-fee-proxy-contract', ()
...
     it('cannot createCreationAction with payment address not an ethereum address', () => {
-      // 'must throw'
       expect(() => {
-        anyToErc20Proxy.createCreationAction({
+        new anyToErc20ProxyPaymentNetwork().createCreationAction({
           paymentAddress: 'not an ethereum address',
           refundAddress: '0x0000000000000000000000000000000000000002',
           salt: 'ea3bc7caf64110ca',
+          network: 'private',
+          acceptedTokens: [DataConversionERC20FeeCreate.tokenAddress],
```
Comment (yomarion) [inline packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts:73]: In order to focus on one error, we need to make everything else correct. Network & acceptedTokens are mandatory. (same comment for many tests below)

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 4]
```diff
@@ -48,397 +28,163 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
+   *
+   * @returns state of the extension created
+   */
+  protected applyCreation(
+    extensionAction: ExtensionTypes.IAction,
+    timestamp: number,
+  ): ExtensionTypes.IState {
+    if (!extensionAction.parameters.network || extensionAction.parameters.network.length === 0) {
+      throw Error('network is required');
+    }
```
Comment (yomarion) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:109]: This is a big change, approve carefully.

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 5]
```diff
@@ -48,397 +28,167 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
-      actionSigner,
-      timestamp,
-    );
-
-    return copiedExtensionState;
-  }
+    const network = creationParameters.network;
+    if (!network) {
+      throw Error('network is required');
+    }
```
Comment (yomarion) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:58]: Approve carefully, this is an important change

File: packages/advanced-logic/test/utils/payment-network/erc20/any-to-erc20-proxy-add-data-generator.ts [hunk 6]
```diff
@@ -13,7 +13,7 @@ export const refundAddress = '0xf17f52151EbEF6C7334FAD080c5704D77216b732';
 export const feeAddress = '0xC5fdf4076b8F3A5357c5E395ab970B5B54098Fef';
 export const feeAmount = '2000000000000000000';
 export const invalidAddress = '0x not and address';
-export const tokenAddress = '0x6b175474e89094c44da98b954eedeac495271d0f';
+export const tokenAddress = '0x9FBDa871d559710256a2502A2517b794B482Db40';
```
Comment (yomarion) [inline packages/advanced-logic/test/utils/payment-network/erc20/any-to-erc20-proxy-add-data-generator.ts:16]: I think there was a typo and I don't know how previous tests let it pass!

File: packages/advanced-logic/test/utils/payment-network/erc20/any-to-erc20-proxy-create-data-generator.ts [hunk 7]
```diff
@@ -99,13 +105,19 @@ export const extensionStateCreatedEmpty = {
     events: [
       {
         name: 'create',
-        parameters: {},
+        parameters: {
+          network: 'private',
+          acceptedTokens: [tokenAddress],
+        },
```
Comment (yomarion) [inline packages/advanced-logic/test/utils/payment-network/erc20/any-to-erc20-proxy-create-data-generator.ts:111]: Previous version + tests made it possible to have no acceptedTokens in these tests, although it is defined as mandatory.

File: packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts (hunk 3 above)
Comment (benjlevesque) [reply packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts:73]: very good point!

File: packages/advanced-logic/src/extensions/payment-network/erc20/fee-proxy-contract.ts [hunk 8]
```diff
@@ -371,3 +371,371 @@ function isValidAddress(address: string): boolean {
 }
 
 export default erc20FeeProxyContract;
+
+export class erc20FeeProxyPaymentNetwork {
```
Comment (benjlevesque) [inline packages/advanced-logic/src/extensions/payment-network/erc20/fee-proxy-contract.ts:375]: a very common convention is to start class names with a capital letter.

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 9]
```diff
@@ -48,397 +28,167 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
+   * @returns state of the request updated
+   */
+  public applyActionToExtension(
+    extensionsState: RequestLogicTypes.IExtensionStates,
+    extensionAction: ExtensionTypes.IAction,
+    requestState: RequestLogicTypes.IRequest,
+    actionSigner: IdentityTypes.IIdentity,
+    timestamp: number,
+  ): RequestLogicTypes.IExtensionStates {
+    return super.applyActionToExtension(
```
Comment (benjlevesque) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:90]: if the only thing a child method does is call its parent, then maybe it can be removed?

File: packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts [hunk 10]
```diff
@@ -12,7 +12,7 @@ describe('extensions/payment-network/erc20/any-to-erc20-fee-proxy-contract', ()
     it('can create a create action with all parameters', () => {
       // 'extension data is wrong'
       expect(
-        anyToErc20Proxy.createCreationAction({
+        new anyToErc20ProxyPaymentNetwork().createCreationAction({
```
Comment (benjlevesque) [inline packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts:15]: why not use the exported default `conversionErc20FeeProxyContract`? I think it would avoid all these changes to the tests

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 11]
```diff
@@ -48,397 +28,167 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
-  timestamp: number,
-): RequestLogicTypes.IExtensionStates {
-  checkSupportedCurrency(requestState.currency, extensionAction.parameters.network || 'rinkeby');
-
-  const copiedExtensionState: RequestLogicTypes.IExtensionStates = Utils.deepCopy(extensionsState);
-
-  if (extensionAction.action === ExtensionTypes.PnFeeReferenceBased.ACTION.CREATE) {
-    if (requestState.extensions[extensionAction.id]) {
-      throw Error(`This extension has already been created`);
+export class anyToErc20ProxyPaymentNetwork extends erc20FeeProxyPaymentNetwork {
```
Comment (benjlevesque) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:31]: should probably implement `IAnyToERC20`

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 12]
```diff
@@ -48,397 +28,167 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
-    );
+    if (!supportedCurrencies[network][request.currency.type].includes(request.currency.value)) {
+      throw new Error(
+        `The currency (${request.currency.value}) of the request is not supported for this payment network.`,
+      );
+    }
   }
 }
 
+const conversionErc20FeeProxyContract: ExtensionTypes.PnAnyToErc20.IAnyToERC20 = new anyToErc20ProxyPaymentNetwork() as ExtensionTypes.PnAnyToErc20.IAnyToERC20;
```
Comment (benjlevesque) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:192]: this part is not required if the class implements the interface
```suggestion
const conversionErc20FeeProxyContract: ExtensionTypes.PnAnyToErc20.IAnyToERC20 = new anyToErc20ProxyPaymentNetwork();
```

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts [hunk 13]
```diff
@@ -48,397 +28,167 @@ const supportedCurrencies: Record<string,  Record<RequestLogicTypes.CURRENCY, st
...
-  const copiedExtensionState: RequestLogicTypes.IExtensionStates = Utils.deepCopy(extensionsState);
-
-  if (extensionAction.action === ExtensionTypes.PnFeeReferenceBased.ACTION.CREATE) {
-    if (requestState.extensions[extensionAction.id]) {
-      throw Error(`This extension has already been created`);
+export class anyToErc20ProxyPaymentNetwork extends erc20FeeProxyPaymentNetwork {
+  public constructor() {
+    super();
+    this.currentVersion = CURRENT_VERSION;
+    this.paymentNetworkId = ExtensionTypes.ID.PAYMENT_NETWORK_ANY_TO_ERC20_PROXY;
```
Comment (benjlevesque) [inline packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:35]: these should be parameters of the constructor

File: packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts (hunk 1 above)
Comment (benjlevesque) [reply packages/advanced-logic/src/extensions/payment-network/any-to-erc20-proxy.ts:44]: I think child methods should have comments. Better for IDE help at least
